doc/neps/ufunc-overrides.rst

```diff
@@ -106,163 +116,193 @@ Returning ``NotImplemented`` to user should not happen. Moreover::
 Here, it appears that the sparse matrix was converted to an object array
 scalar, which was then multiplied with all elements of the ``b`` array.
 However, this behavior is more confusing than useful, and having a
-``TypeError`` would be preferable.
+:exc:`TypeError` would be preferable.
 
-Adding the ``__numpy_ufunc__`` functionality fixes this and would
+Adding the ``__array_ufunc__`` functionality fixes this and would
 deprecate the other ufunc modifying functions.
 
 .. [5] http://mail.python.org/pipermail/numpy-discussion/2011-June/056945.html
 
+.. [6] https://github.com/numpy/numpy/issues/5844
 
 Proposed interface
 ==================
 
-Objects that want to override Ufuncs can define a ``__numpy_ufunc__`` method.
-The method signature is::
+The standard array class :class:`ndarray` gains an ``__array_ufunc__``
+method and objects can override Ufuncs by overriding this method (if
+they are :class:`ndarray` subclasses) or defining their own. The method
+signature is::
 
-    def __numpy_ufunc__(self, ufunc, method, i, inputs, **kwargs)
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs)
 
 Here:
 
 - *ufunc* is the ufunc object that was called. 
 - *method* is a string indicating which Ufunc method was called
   (one of ``"__call__"``, ``"reduce"``, ``"reduceat"``,
   ``"accumulate"``, ``"outer"``, ``"inner"``). 
-- *i* is the index of *self* in *inputs*.
 - *inputs* is a tuple of the input arguments to the ``ufunc``
 - *kwargs* are the keyword arguments passed to the function. The ``out``
-  arguments are always contained in *kwargs*, how positional variables
-  are passed is discussed below.
-
-The ufunc's arguments are first normalized into a tuple of input data
-(``inputs``), and dict of keyword arguments. If there are output
-arguments they are handled as follows:
-
-- One positional output variable x is passed in the kwargs dict as ``out :
-  x``.
-- Multiple positional output variables ``x0, x1, ...`` are passed as a tuple
-  in the kwargs dict as ``out : (x0, x1, ...)``.
-- Keyword output variables like ``out = x`` and ``out = (x0, x1, ...)`` are
-  passed unchanged to the kwargs dict like ``out : x`` and ``out : (x0, x1,
-  ...)`` respectively.
-- Combinations of positional and keyword output variables are not
-  supported.
+  arguments are always contained as a tuple in *kwargs*.
+
+Hence, the arguments are normalized: only the input data (``inputs``)
+are passed on as positional arguments, all the others are passed on as a
+dict of keyword arguments (``kwargs``). In particular, if there are
+output arguments, positional are otherwise, they are passed on as a
+tuple in the ``out`` keyword argument.
```

Comment:
@eric-wieser - you'd have to get `out` out of `kwargs` anyway, so your example would just be:
```
for o in kwargs.get('out', ()):
   ...
```
now in reality you will usually need special handling twice, once before actually doing anything with the inputs, to be sure the output can receive it, and another time after, in recasting results. I have found myself doing something like
```
outputs = kwargs.get('out', None)
if outputs is not None:
    ...
```
I actually think this makes the code path clearer -- no outputs given, no need to look in this part.

As another argument: why all this focus on `out` specifically? There are other optional arguments too: `dtype`, `axis`, `subok`, `where`, `signature`, `casting`, `order`, `extobj`. There is an advantage to knowing that "all is default" by having an empty `kwargs`.

Indeed, we never really decided whether we needed to look on `where` for `__array_ufunc__` as well...